### PR TITLE
Add AppCR config version handler

### PR DESCRIPTION
**Pull request by kubasobon**
## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: service/controller/app.go [hunk 1]
```diff
@@ -91,6 +107,7 @@ func newAppResources(config AppConfig) ([]resource.Interface, error) {
 	}
 
 	resources := []resource.Interface{
+		configversionResource,
```
Comment (kopiczko) [inline service/controller/app.go:110]: Even though it doesn't make much sense now I'd move it to a separate controller (app_catalog_entry.go) to make our life easier when we are unblocked with https://github.com/giantswarm/giantswarm/issues/14832.

File: service/controller/handler/configversion/create.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	h.logger.Debugf(ctx, "setting App %#q config version", app.Spec.Name)
+
+	h.logger.Debugf(ctx, "resolving config version for App %#q from %#q catalog", app.Name, app.Spec.Catalog)
+	var configVersion string
+
+	var index Index
+	{
+		store, err := h.gitHub.GetFilesByBranch(ctx, owner, app.Spec.Catalog, "master")
```
Comment (kopiczko) [inline service/controller/handler/configversion/create.go:45]: I think it's simpler to do http get this https://giantswarm.github.io/control-plane-catalog/index.yaml

File: service/controller/handler/configversion/handler.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	AppVersion    string   `json:"appVersion"`
+	ConfigVersion string   `json:"configVersion,omitempty"`
+	Created       string   `json:"created"`
+	Description   string   `json:"description"`
+	Digest        string   `json:"digest"`
+	Home          string   `json:"home"`
+	Name          string   `json:"name"`
+	Urls          []string `json:"urls"`
+	Version       string   `json:"version"`
+}
```
Comment (kopiczko) [inline service/controller/handler/configversion/handler.go:77]: Usually we put data types in `types.go`. Also: do we really need so many fields?

File: service/controller/handler/configversion/handler.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/giantswarm/k8sclient/v5/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	"github.com/giantswarm/config-controller/service/internal/github"
+)
+
+const (
+	Name  = "configversion"
+	owner = "giantswarm"
```
Comment (kopiczko) [inline service/controller/handler/configversion/handler.go:13]: I think it's used once, so maybe there is no point in making it a const.

File: service/controller/handler/configversion/create.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
+package configversion
+
+import (
+	"context"
+
+	"github.com/ghodss/yaml"
+	"github.com/giantswarm/apiextensions/v3/pkg/annotation"
+	"github.com/giantswarm/microerror"
+
+	controllerkey "github.com/giantswarm/config-controller/service/controller/key"
```
Comment (kopiczko) [inline service/controller/handler/configversion/create.go:10]: Why aliasing this import?

File: service/controller/handler/configversion/create.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+func (h *Handler) EnsureCreated(ctx context.Context, obj interface{}) error {
+	app, err := controllerkey.ToAppCR(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	value, ok := app.GetAnnotations()[annotation.ConfigVersion]
+	if ok {
+		h.logger.Debugf(ctx, "App CR %#q is already annotated with %#q: %s", app.Name, annotation.ConfigVersion, value)
```
Comment (kopiczko) [inline service/controller/handler/configversion/create.go:21]: I think the value should be also reconciled. That's especially important when app is upgraded to a newer version which may use different configuration version.

Comment (kopiczko) [inline service/controller/handler/configversion/create.go:21]: There is no need to mention app.Name in the logs. There is already `object` key injected into all handler log messages.

File: service/controller/handler/configversion/create.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	entries, ok := index.Entries[app.Spec.Name]
+	if !ok || len(entries) == 0 {
+		h.logger.Debugf(ctx, "App %#q has no entries in %#q's index.yaml", app.Spec.Name, app.Spec.Catalog)
+		h.logger.Debugf(ctx, "cancelling handler")
+		return nil
+	}
+
+	configVersion = entries[0].ConfigVersion
```
Comment (kopiczko) [inline service/controller/handler/configversion/create.go:68]: Entry selection should be based on matching version. Different versions of an app may require different config version.

File: service/controller/handler/configversion/create.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	if configVersion == "" {
+		configVersion = "0.0.0"
+	}
+	h.logger.Debugf(ctx, "resolved config version for App %#q from %#q catalog", app.Name, app.Spec.Catalog)
+
+	annotations := app.GetAnnotations()
+	annotations[annotation.ConfigVersion] = configVersion
```
Comment (kopiczko) [inline service/controller/handler/configversion/create.go:75]: I think it may happen that annotations are nil. The map should be initialized if it's nil just in case.

File: service/controller/handler/configversion/create.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	if app.Spec.Catalog == "" {
+		h.logger.Debugf(ctx, "App CR %#q has no .Spec.Catalog set", app.Name)
+		h.logger.Debugf(ctx, "cancelling handler")
+		return nil
+	}
+
+	if app.Spec.Catalog == "releases" {
+		h.logger.Debugf(ctx, "App CR %#q has a \"releases\" catalog set", app.Name)
```
Comment (kopiczko) [inline service/controller/handler/configversion/create.go:33]: When is the pause annotation removed in that case?

File: service/controller/handler/values/create.go [hunk 10]
```diff
@@ -20,7 +20,7 @@ func (h *Handler) EnsureCreated(ctx context.Context, obj interface{}) error {
 
 	configVersion, ok := app.GetAnnotations()[annotation.ConfigVersion]
 	if !ok {
-		h.logger.Debugf(ctx, "App CR %q is missing %q annotation", app.Name, annotation.ConfigVersion)
+		h.logger.Debugf(ctx, "App CR %#q is missing %#q annotation", app.Name, annotation.ConfigVersion)
 		if _, ok := app.GetAnnotations()[PauseAnnotation]; ok {
 			err = h.removeAnnotation(ctx, &app, PauseAnnotation)
```
Comment (kopiczko) [inline service/controller/handler/values/create.go:25]: It shouldn't be removed here IMO. This handler should wait till it's set. It's set on all apps except those from releases catalog (which need a special handling).

File: service/controller/app_catalog_entry.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+			K8sClient: config.K8sClient,
+			Logger:    config.Logger,
+			NewRuntimeObjectFunc: func() runtime.Object {
+				return new(v1alpha1.App)
+			},
+			Resources: resources,
+			Selector:  label.AppVersionSelector(config.UniqueApp),
+
+			// Name is used to compute finalizer names. This here results in something
+			// like operatorkit.giantswarm.io/config-controller-app-controller.
```
Comment (JosephSalisbury) [inline service/controller/app_catalog_entry.go:51]: I doubt this comment

Comment (kubasobon) [reply service/controller/app_catalog_entry.go:51]: I took it from here: https://github.com/giantswarm/template-operator/blob/master/service/controller/todo.go#L45-L46
What's wrong with it?

File: service/controller/handler/configversion/create.go (hunk 6 above)
Comment (kubasobon) [reply service/controller/handler/configversion/create.go]: Agreed.

File: service/controller/handler/configversion/create.go (hunk 9 above)
Comment (kubasobon) [reply service/controller/handler/configversion/create.go]: I moved annotation removal around a bit. Now it's removed from releases here and from the other resources once the config version is set and config generated.

File: service/controller/handler/values/create.go [hunk 12]
```diff
@@ -9,34 +9,38 @@ import (
...
-			if err != nil {
-				return err
-			}
-		}
+		h.logger.Debugf(ctx, "App CR is missing %q annotation", annotation.ConfigVersion)
+		h.logger.Debugf(ctx, "cancelling handler")
+		return nil
+	}
+
+	if configVersion == "0.0.0" {
```
Comment (kopiczko) [inline service/controller/handler/values/create.go:32]: It feels like `0.0.0` should be extracted as a named const in the `key` package.

File: service/controller/handler/configversion/create.go [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+			h.logger.Debugf(ctx, "cancelling handler")
+			return nil
+		}
+	}
+
+	annotations[annotation.ConfigVersion] = configVersion
+	if configVersion == "0.0.0" {
+		h.logger.Debugf(ctx, "App does not use generated config, removing pause annotation")
+		annotations = key.RemoveAnnotation(annotations, key.PauseAnnotation)
+	}
```
Comment (kopiczko) [inline service/controller/handler/configversion/create.go:101]: I'd say that this should be `values` handler responsibility. Otherwise we'll have the same `if` in multiple places.

Comment (kubasobon) [reply service/controller/handler/configversion/create.go]: Agreed, I moved the code to `values`.

File: service/controller/handler/values/create.go (hunk 12 above)
Comment (kubasobon) [reply service/controller/handler/values/create.go]: True. Let me know if the name sound ok to you.

File: service/controller/app_catalog_entry.go (hunk 11 above)
Comment (JosephSalisbury) [reply service/controller/app_catalog_entry.go:51]: oh, i think i read the 'like' here wrong - i'd expect this to end up with `operatorkit.giantswarm.io/config-controller-app-catalog-entry-controller.`

tl;dr i'm a massive nit picker

File: service/controller/key/key.go [hunk 14]
```diff
@@ -7,6 +7,16 @@ import (
 	"github.com/giantswarm/microerror"
 )
 
+const (
+	// PauseAnnotation stops app-operator from reconciling App CR too soon.
+	// Once config has been generated, this annotation should be removed by
+	// config-controller.
+	PauseAnnotation = "app-operator.giantswarm.io/paused"
```
Comment (kopiczko) [inline service/controller/key/key.go:14]: We should use const from apiextensions instead.

**Review by kopiczko — APPROVED** [review]

**Comment by kubasobon** [comment]
Tested on _geckon_ with https://github.com/giantswarm/app-admission-controller/pull/61.
